File: Java/AutomationFramework/src/com/app/helpers/actions/DriverActions.java
package com.app.helpers.actions;

import com.app.helpers.logging.ErrorLogger;
import com.app.helpers.logging.ILogger;
import com.app.helpers.logging.InfoLogger;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import java.time.Duration;
import java.util.List;

public class DriverActions {
    public WebDriver driver;
    public MouseActions mouseActions;
    public KeyboardActions keyBoardActions;
    public WindowActions windowActions;

    public DriverActions(WebDriver driver)
    {
        this.driver = driver;
        mouseActions = new MouseActions(driver);
        keyBoardActions = new KeyboardActions(driver);
        windowActions = new WindowActions(driver);
    }

    public DriverActions() {

    }

    public String GetText(WebElement element)
    {
        return WaitUntilElementVisibleAndClickable(element).getText();
    }

    public String GetAttributes(WebElement element, String attribute)
    {
        return WaitUntilElementVisibleAndClickable(element).getAttribute(attribute);
    }

    public boolean GetPresenceOfElement(WebElement element)
    {
        boolean result = WaitUntilElementVisibleAndClickable(element).isDisplayed();
        ILogger.PrintLog(new InfoLogger().LogMessage("Element '"+ element + "' is present"));
        return result;
    }

    public boolean GetVisibilityOfElement(WebElement element)
    {
        boolean result = false;
        try
        {
            result = WaitUntilElementVisibleAndClickable(element).isDisplayed();
        }
        catch(WebDriverException exp)
        {
            ILogger.PrintLog(new ErrorLogger().LogMessage("Element '"+ element + "' is not present"));
        }
        return result;
    }

    public List<String> GetText(List<WebElement> elements)
    {
        List<String> list = elements.stream().map(elt -> elt.getText()).toList();
        return list;
    }

    public WebElement WaitUntilElementVisibleAndClickable(WebElement element)
    {
        try {
            element = new WebDriverWait(driver, Duration.ofSeconds(15))
                    .until(ExpectedConditions.elementToBeClickable(element));
            ILogger.PrintLog(new InfoLogger().LogMessage("Element '"+ element + "' displayed"));
        }
        catch (TimeoutException exp) {
            ILogger.PrintLog(new ErrorLogger().LogMessage("Element '"+ element + "' not displayed"));
        }
        return element;
    }
}

File: DotNet/Framework/AutomationFramework/Helpers/Actions/DriverActions.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using Helpers.Logging;
using SeleniumExtras.WaitHelpers;

namespace Helpers.Actions
{
    public class DriverActions
    {
        public IWebDriver driver { get; set; }
        public MouseActions mouseActions;
        public KeyBoardActions keyBoardActions;
        public WindowActions windowActions;

        public DriverActions(IWebDriver driver)
        {
            this.driver = driver;
            mouseActions = new MouseActions(driver);
            keyBoardActions = new KeyBoardActions(driver);
            windowActions = new WindowActions(driver);
        }

        public string GetText(By selector)
        {
            return WaitUntilElementDisplayed(selector).Text;
        }

        public string GetAttributes(By selector, string attribute)
        {
            return WaitUntilElementDisplayed(selector).GetAttribute(attribute);
        }

        public bool GetPresenceOfElement(By selector)
        {
            bool result = WaitUntilElementDisplayed(selector).Displayed;
            Logger.PrintLog(new InfoLogger().LogMessage($"Element '{selector}' is present"));
            return result;
        }

        public bool GetVisibilityOfElement(By selector)
        {
            bool result = false;
            try
            {
                result = WaitUntilElementDisplayed(selector).Displayed;
                Logger.PrintLog(new InfoLogger().LogMessage($"Element '{selector}' is present"));
            }
            catch(WebDriverTimeoutException exp)
            {
                Logger.PrintLog(new ErrorLogger().LogMessage($"Element '{selector}' is not present"));
            }
            return result;
        }

        public void GetText(By selector, out List<string> list)
        {
            WaitUntilElementDisplayed(selector);
            list = driver.FindElements(selector)
                .Select(e => e.Text).ToList();
        }

        public void GetAttributes(By selector, string attribute, out List<string> list)
        {
            WaitUntilElementDisplayed(selector);
            list = driver.FindElements(selector)
                .Select(e => e.GetAttribute(attribute)).ToList();
        }

        public virtual IWebElement WaitUntilElementDisplayed(By selector)
        {
            WebDriverWait wait = new(driver, new TimeSpan(0, 0, 15));
            IWebElement element = wait.Until(ExpectedConditions.ElementIsVisible(selector));
            Logger.PrintLog(new InfoLogger().LogMessage($"Element '{selector}' displayed"));
            return element;
        }

        public virtual IWebElement WaitUntilElementClickable(By selector)
        {
            WebDriverWait wait = new(driver, new TimeSpan(0, 0, 15));
            IWebElement element = wait.Until(ExpectedConditions.ElementToBeClickable(selector));
            Logger.PrintLog(new InfoLogger().LogMessage($"Element '{selector}' displayed"));
            return element;
        }
    }
}
